### PR TITLE
Enable .NET 6 build [API-1028]

### PR DESCRIPTION
**Pull request by zpqrtbnk**
.NET 6 has been released in November 2021 and is now the official long-term supported .NET release. It brings various performance improvements, esp. around networking. The previous .NET 5 version is *not* LTS, and will not be supported beyond May 2022. The previous LTS version is 3.1 which was released in December 2019, it is missing all the important performance improvements that have been added since that date, and it will not be supported beyond December 2022.

It is therefore important that we support the latest, officially supported version of .NET.

This PR:
* Fixes the `hz.ps1` script for better handling of frameworks, including the new `net6.0`.
* Fixes for some known issues in .NET 6 around async (see note below).

Review:
* All GitHub workflows have been modified to install .NET 6.
* All `.csproj` files have been modified to add support for .NET 6.
* A range of `.cs` code files have been modified to deal with the APM/TAP async issues

Note: some old APM (the "old way") networking code has documented issues in .NET 6, and noone even at MS seems to figure out how to fix them, and the recommendation is to use TAP (the "new" way). However, since TAP is not available for older versions, we have to use `#ifdef` blocks.

More details: this happens in `SocketExtensions.cs`:

* The APM (Asynchronous Programming Model) model is implemented in `ConnectAsync4` and consists in invoking `socket.BeginConnect(...)` passing a callback method, which in turn will invoke `socket.EndConnect(...)`. 
* That is how async was done previously, and in theory this way of doing things is supported and *should work* in .NET 5 or 6. However, with .NET 6 it leaves some unobserved exceptions that can render the application unstable. This is [this .NET issue](https://github.com/dotnet/runtime/issues/61411) which is 4+ months old and still open. And, judging by the internal code, not an easy fix.
* The alternative is to switch over to the TAP (Task-based Asynchronous Pattern) model which consists in invoking one single `socket.ConnectAsync(...)` mehod. It's cleaner, but is not available before .NET 5. It is implemented in `ConnectAsync5`.

Conclusion: my original idea was to stick with the original code to reduce changes, but it's not stable on .NET 6, and the new code is not available before .NET 5, so we end up needing two different implementations.

### CODE REVIEW

File: src/Hazelcast.Net.Tests/DotNet/ExceptionTests.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using NUnit.Framework;
+using Hazelcast.Core;
+
+namespace Hazelcast.Tests.DotNet
+{
+    [TestFixture]
+    public class ExceptionTests
+    {
+        [Test]
+        public void MustThrowToGetStackTrace()
```
Comment (emreyigit) [inline src/Hazelcast.Net.Tests/DotNet/ExceptionTests.cs:26]: What is the goal here? Do we expect something different between .net versions? What am I missing?

**Review by emreyigit — APPROVED** [review]

**Comment by ihsandemir** [comment]
Can you please share details about these issues mentioned?: `some old APM (the "old way") networking code has documented issues in .NET 6`

**Comment by ihsandemir** [comment]
Can you rebase the PR?

**Comment by zpqrtbnk** [comment]
Have rebased the PR + updated the description with more details re. async issues, including link to the MS issue.

File: src/Hazelcast.Net/Core/SocketExtensions.cs [hunk 2]
```diff
@@ -27,22 +27,98 @@ namespace Hazelcast.Core
...
+
+            if (cancellationToken.IsCancellationRequested)
+                throw new OperationCanceledException("The socket connection operation has been canceled.");
+
+            throw new TimeoutException("The socket connection operation has timed out.");
+        }
+
+        public static async Task ConnectAsync4(this Socket socket, EndPoint endPoint, int timeoutMilliseconds, CancellationToken cancellationToken)
+        {
+            // note that purely using this code in .NET 6 leaves an unobserved exception behind
```
Comment (ihsandemir) [inline src/Hazelcast.Net/Core/SocketExtensions.cs:120]: can you put an assert statement to guarantee this?

File: hz.ps1 [hunk 3]
```diff
@@ -889,46 +930,42 @@ function require-dotnet ( $full ) {
...
-        Write-Output ""
-        Write-Output "This script requires Microsoft .NET Core 5.0.200+ SDK, which can be downloaded at: https://dotnet.microsoft.com/download/dotnet-core"
-        Die "Could not find dotnet SDK version 5.0.200+"
+    # validate that we have the required SDKs based upon the complete list of
+    # frameworks for the test project, as determined by determine-target-frameworks.
+
+    $result = @{ validSdks = $true; sdkInfos = "  SDKs:" }
+    require-dotnet-version $result $sdks "2.1" $frameworks "netcoreapp2.1" "Core 2.1.x" $allowPrerelease
+    require-dotnet-version $result $sdks "3.1" $frameworks "netcoreapp3.1" "Core 3.1.x" $allowPrerelease
+    require-dotnet-version $result $sdks "5.0" $frameworks "net5.0" "5.0.x" $allowPrerelease
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply hz.ps1:947]: See added line 958 below `require-dotnet-version $result $sdks "6.0" $frameworks "net6.0" "6.0.x" $allowPrerelease`

File: src/Hazelcast.Net/Core/ExceptionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        /// </summary>
+        /// <typeparam name="TException">The type of the exception.</typeparam>
+        /// <param name="exception">The exception.</param>
+        /// <returns>The exception after it has been thrown and caught.</returns>
+        /// <remarks>
+        /// <para>This can be useful when creating, but not immediately throwing, an exception, such
+        /// as with <code>taskContinuationSource.TrySetException(new Exception())</code> in order
+        /// to force the new exception to have a stack trace that corresponds to its creation.</para>
+        /// </remarks>
+        public static TException Thrown<TException>(this TException exception)
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net/Core/ExceptionExtensions.cs:36]: Similar but different.

File: src/Hazelcast.Net.Tests/Messaging/ClientMessageConnectionTests.cs [hunk 5]
```diff
@@ -140,6 +140,7 @@ static ClientMessage NewFragment(ClientMessageFlags flags, long fragmentId = 0)
         }
 
         [Test]
+        [Timeout(20_000)]
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Tests/Messaging/ClientMessageConnectionTests.cs:143]: While fixing the issues with .NET 6 and async sockets, this test could run forever. This makes sure it fails after a timeout.

File: src/Hazelcast.Net/Core/SocketExtensions.cs (hunk 2 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net/Core/SocketExtensions.cs:120]: I don't understand what you want to assert. Let's discuss it.

File: src/Hazelcast.Net/Core/ExceptionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Hazelcast.Core
+{
+    /// <summary>
+    /// Provides extension methods for exceptions.
+    /// </summary>
+    internal static class ExceptionExtensions
+    {
+        /// <summary>
+        /// Throws and catches an exception, thus forcing it to have a stack trace.
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net/Core/ExceptionExtensions.cs:26]: documenting.

File: src/Hazelcast.Net.Tests/Networking/NetworkingTests.cs [hunk 7]
```diff
@@ -613,6 +614,204 @@ public void Sequences1()
             NUnit.Framework.Assert.AreEqual(origin, value);
         }
 
+        [Test]
+        [Timeout(20_000)]
+        [Repeat(2)]
+        public async Task Net6Repro()
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Tests/Networking/NetworkingTests.cs:620]: documenting.

File: hz.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (emreyigit) [reply hz.ps1:947]: It is at line 958

**Comment by zpqrtbnk** [comment]
FWIW I have updated the PR, made changes to the tests etc so it should be way cleaner now.

**Review by emreyigit — APPROVED** [review]

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**